driver/labjack/writer_sink.cpp

```diff
@@ -0,0 +1,262 @@
+
+    this->get_index_keys(); // retrieve state index from first state channel
+
+    this->state_source = std::make_shared<labjack::StateSource>(
+        this->writer_config.state_rate,
+        this->writer_config.state_index_key,
+        this->writer_config.initial_state_map
+    );
+}
+
+labjack::WriteSink::~WriteSink() {
+    this->stop("");
+}
+
+void labjack::WriteSink::init() {
+    if (this->writer_config.device_type == "") {
+        auto [dev, err] = this->ctx->client->hardware.retrieveDevice(
+            this->writer_config.device_key
+        );
+        if (err != freighter::NIL) {
+            LOG(ERROR) << "[labjack.writer] Error retrieving device: " << err.message();
+            return;
+        }
+        this->writer_config.device_type = dev.model;
+    } {
+        std::lock_guard<std::mutex> lock(labjack::device_mutex);
+        check_err(LJM_Open(LJM_dtANY, LJM_ctANY,
+                           this->writer_config.serial_number.c_str(), &this->handle));
+    }
+}
+
+freighter::Error labjack::WriteSink::write(synnax::Frame frame) {
+    auto frame_index = 0;
+    for (auto key: *(frame.channels)) {
+        double value = series_to_val(frame.series->at(frame_index));
+        std::string loc = this->writer_config.initial_state_map[key].location;
+        check_err(LJM_eWriteName(this->handle, loc.c_str(), value));
+        frame_index++;
+    }
+    this->state_source->update_state(std::move(frame));
+    return freighter::NIL;
+}
+
+freighter::Error labjack::WriteSink::stop(const std::string &cmd_key) {
+    ctx->setState({
+        .task = task.key,
+        .key = cmd_key,
+        .variant = "success",
+        .details = {
+            {"running", false},
+            {"message", "Task stopped successfully"}
+        }
+    });
+    return freighter::NIL;
+}
+
+freighter::Error labjack::WriteSink::start(const std::string &cmd_key) {
+    this->init();
+    ctx->setState({
+        .task = task.key,
+        .key = cmd_key,
+        .variant = "success",
+        .details = {
+            {"running", true},
+            {"message", "Task started successfully"}
+        }
+    });
+    return freighter::NIL;
+}
+
+std::vector<synnax::ChannelKey> labjack::WriteSink::get_cmd_channel_keys() {
+    std::vector<synnax::ChannelKey> keys;
+    for (auto &channel: this->writer_config.channels) {
+        if (channel.enabled) keys.push_back(channel.cmd_key);
+    }
+    // Don't need index key as we're only using this for streaming cmds
+    return keys;
+}
+
+std::vector<synnax::ChannelKey> labjack::WriteSink::get_state_channel_keys() {
+    std::vector<synnax::ChannelKey> keys;
+    for (auto &channel: this->writer_config.channels) {
+        if (channel.enabled) keys.push_back(channel.state_key);
+    }
+    keys.push_back(this->writer_config.state_index_key);
+    return keys;
+}
+
+void labjack::WriteSink::get_index_keys() {
+    if (this->writer_config.channels.empty()) {
```

Comment:
A getter should always return something IMo. Also, no guarantee that there is only one state index key.

Reply:
oh yeah i had a comment here to change that i missed